### PR TITLE
CCM-6319: initial framework, pipeline job and tests

### PR DESCRIPTION
**Pull request by sean-corrigan1**
## Description

Create initial framework for running automated UI component tests:
- Set up playwright dependencies
- Create initial config to allow local running of tests
- Create setup / teardown of session data to allow seeding of data in Amplifly dynamoDB
- Create Page Object Models for initial components (pages)
- Create initial tests for components (pages)
- Integrate UI component tests into CI/CD Github workflow
- Output test report as artefact with screenshots (on failure) from CI/CD Github workflow

## Context

To automate UI testing and minimise manual test effor

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [x] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,6 +10,8 @@
     "lint:fix": "next lint  --dir . --fix",
     "test:unit": "jest",
     "test:accessibility": "pa11y-ci -c ./tests/accessibility/.pa11y-ci.js",
+    "test:local-ui": "npx playwright test --project component -c tests/test-team/config/local.config.ts",
```
Comment (mark-r-bjss) [inline package.json:13]: Should these be their own project/package.json - given that you can see some application errors during tests?

File: tests/test-team/pages/template-mgmt-choose-page.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  constructor(page: Page) {
+    super(page);
+    this.chooseTemplateRadioGroup = page.getByRole('radiogroup');
+    this.fieldsetHeading = page.locator('[class="nhsuk-fieldset__heading"]');
+  }
+
+  static emptySessionData: Session = {
+    id: '3d98b0c4-6666-0000-0000-95eb27590000',
+    templateType: 'UNKNOWN',
+    nhsAppTemplateName: ' ',
```
Comment (mark-r-bjss) [inline tests/test-team/pages/template-mgmt-choose-page.ts:19]: Shouldn't this be an empty string rather than a space?

File: tests/test-team/pages/template-mgmt-create-page.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// import { type Page } from '@playwright/test';
```
Comment (mark-r-bjss) [inline tests/test-team/pages/template-mgmt-create-page.ts:1]: Can you check that the linter and typechecker are covering these files.  The linter should have picked up on these comments.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  test('should display NHS App radio button selected if present in session storage', async ({
+    page,
+  }) => {
+    const chooseTemplatePage = new TemplateMgmtChoosePage(page);
+
+    await chooseTemplatePage.navigateToChooseTemplatePage(
+      TemplateMgmtChoosePage.nhsAppRadioSelectedSessionData.id
+    );
+
+    // expect to assert radio selected
```
Comment (mark-r-bjss) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts:123]: Is this deliberate?

File: tests/test-team/setup/session.setup.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    ...TemplateMgmtChoosePage.sessionData,
+    ...TemplateMgmtCreatePage.sessionData,
+  ];
+
+  Amplify.configure(config);
+
+  const client = generateClient<Schema>({ authMode: 'iam' });
+
+  for (const data of sessionData) {
+    await client.models.SessionStorage.create(data);
```
Comment (mark-r-bjss) [inline tests/test-team/setup/session.setup.ts:27]: The benefit of this is that it is quick to develop now but the downside is that the test is not independent of the application code because it is using the graphQL API to create the data.  An alternative would be to go directly to the database via an independent auth role.  I'm not sure how this approach will work when we have a more mature auth layer as we won't have guest access to the APIs.

File: tests/test-team/pages/template-mgmt-base-page.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  readonly goBackLink: Locator;
+
+  readonly pageHeader: Locator;
+
+  readonly continueButton: Locator;
+
+  constructor(page: Page) {
+    this.page = page;
+    this.notifyBannerLink = page.locator(
+      '[class="nhsuk-header__link nhsuk-header__link--service"]'
```
Comment (bhansell1) [inline tests/test-team/pages/template-mgmt-base-page.ts:19]: What do you think about using a `data-*` attribute on HTML elements on the code instead of relying on `classnames` and `text` which are subject to change and in the case of a class return more than 1 object.

For example if we have some html:

```
<button data-test-id="pw-important-buttom">
Click me!
</button>
```

Your locators can use the data-* attribute like:

```
page.locator('data-testid=important-button')
```

I found this [article](https://playwrightsolutions.com/getbytestid/) which seems reasonably interesting and it looks like Playwright [has native support for finding elements by data-* attribute](https://playwrightsolutions.com/getbytestid/).

File: tests/test-team/setup/session.setup.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable no-await-in-loop */
+/* eslint-disable unicorn/prefer-module */
+/* eslint-disable import/no-unresolved */
+/* eslint-disable @typescript-eslint/no-var-requires */
```
Comment (bhansell1) [inline tests/test-team/setup/session.setup.ts:4]: is this needed?

I generally follow the rule of them where I don't inline disable a rule, unless it's a super edge case. If the rule isn't valid or I disagree with it. Then I'd disable it globally or update the code to fix the linting issue. 

I think in the case of the ones you've disabled they might not be relevant any more? Apart from the `no-await-in-loop`.

Which is valid. You can update your code to promisify it like so;

```
const promises = sessionData.map((session) => client.create(session));

await Promise.all(promises);
```

This also has additional benefits of being slightly faster because it's "concurrent". Also if any of the promises fail then it'll throw an error. The alternative is `Promise.allSettled` which doesn't fail but you have to reason over the response.

File: tests/test-team/teardown/session.teardown.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  const sessionData: Session[] = [
+    ...TemplateMgmtChoosePage.sessionData,
+    ...TemplateMgmtCreatePage.sessionData,
+  ];
+
+  Amplify.configure(config);
+
+  const client = generateClient<Schema>({ authMode: 'iam' });
+
+  for await (const data of sessionData) {
```
Comment (bhansell1) [inline tests/test-team/teardown/session.teardown.ts:26]: Same here you can Promisify it. Using Promise.all() or Promise.allSettled.

File: tests/test-team/pages/template-mgmt-choose-page.ts (hunk 2 above)
Comment (sean-corrigan1) [reply tests/test-team/pages/template-mgmt-choose-page.ts]: Yes ... sorted.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts (hunk 4 above)
Comment (sean-corrigan1) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts:123]: Yes, will be picked up by CCM-6321 to complete.

File: tests/test-team/setup/session.setup.ts (hunk 5 above)
Comment (sean-corrigan1) [reply tests/test-team/setup/session.setup.ts]: Updated so as not to use the GraphQL api to create data ... the dynamo table name is ascertained by matching GraphQL api endpoints between that in the amplify outputs and that listed using the a method from AppSync SDK. Then uploaded directly to dynamo.

File: tests/test-team/pages/template-mgmt-base-page.ts (hunk 6 above)
Comment (sean-corrigan1) [reply tests/test-team/pages/template-mgmt-base-page.ts:19]: This was a choice not to use the test id based on what i'm used to, and in my head, the html should be testable without the addition of special test attributes (yes, i disagree with an established test thing ... probably a losing battle). 

The fact they are subject to change makes it more valuable rather than an introduction of flakiness. While we are explicitly testing the functionality of, say a button, the use of class / role and text implicitly tests the button has the correct text and that the button is a button and not a link (both would would pass a functional test as they utilise the .click() method if using a data-testid locator that hasn't been updated).

If the element cant be found then something has gone wrong / changed by accident / unexpectedly, test id may not pick up any of this.

Happy to iterate on it / discuss if my logic sounds to picky / trying to do too much.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      'Choose a template type to create'
+    );
+
+    await expect(page.locator('[class="nhsuk-error-summary"]')).toBeVisible();
+
+    await expect(
+      page
+        .locator('[class="nhsuk-list nhsuk-error-summary__list"]')
+        .getByRole('listitem')
+    ).toHaveText(['Select a template type']);
```
[the remark this answers is not in the document]
Comment (sean-corrigan1) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts:92]: Something to think about, i may just move this to https://nhsd-jira.digital.nhs.uk/browse/CCM-6321 where we will review the current tests 👍 It is something thats dynamic on the page and i believe it modifies the url so maybe something to add in with above ticket and use it as a pattern moving forward.

File: tests/test-team/setup/session.setup.ts (hunk 7 above)
Comment (sean-corrigan1) [reply tests/test-team/setup/session.setup.ts]: This is cool, got a bit stuck with the async/await stuff but this is nice and clean 🦖

File: package.json (hunk 1 above)
Comment (sean-corrigan1) [reply package.json]: Updated to separate out package.json and removed npx bit

File: tests/test-team/pages/template-mgmt-create-page.ts (hunk 3 above)
Comment (sean-corrigan1) [reply tests/test-team/pages/template-mgmt-create-page.ts]: Outdated, commented out code uncommented.

File: tests/test-team/pages/template-mgmt-base-page.ts (hunk 6 above)
Comment (bhansell1) [reply tests/test-team/pages/template-mgmt-base-page.ts:19]: Hey, that's a completely valid reason! 

You're right in saying searching by specifics like by `text` or `role` is a good way of finding elements. 

My one concern is using class names in locators as that can be quite wide ranging. More of a nitpick than anything!

File: tests/test-team/template-mgmt-component-tests/template-mgmt-start-page.component.ts [hunk 10]
```diff
@@ -83,6 +83,33 @@ test.describe('Start Page', () => {
...
+      name: 'Privacy Policy',
+      selector: 'a[data-testid="privacy-policy-link"]',
+    },
+    {
+      name: 'Terms and Conditions',
+      selector: 'a[data-testid="terms-and-conditions-link"]',
+    },
+  ];
+
+  Promise.all(
```
Comment (bhansell1) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-start-page.component.ts:107]: Probably need an await here.

File: tests/test-team/package.json [hunk 11]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "nhs-notify-web-template-management-ui-tests",
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "test:start-local-app": "npm run dev --prefix ../../",
```
Comment (bhansell1) [inline tests/test-team/package.json:6]: This is a really great idea!

I wonder whether we should be running against a production build vs running in dev. OFC running `npm run dev` is faster but doing a `npm run build && npm run start` is more closer aligned with the deployed version of the app. 

There are small differences with how the app works in both modes (I remember some time ago we had really flakey PW tests due to running in dev mode). 

I'll add a comment to one of the tickets you've raised.

File: tests/test-team/setup/session.setup.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import { test as setup } from '@playwright/test';
+import { TemplateMgmtChoosePage } from '../pages/template-mgmt-choose-page';
+import { TemplateMgmtCreatePage } from '../pages/template-mgmt-create-page';
+import { Session } from '../helpers/types';
+import SessionStorageHelper from '../helpers/session-storage-helper';
+
+setup('setup session data', async () => {
+  const sessionData: Session[] = [
+    ...TemplateMgmtChoosePage.sessionData,
```
Comment (bhansell1) [inline tests/test-team/setup/session.setup.ts:9]: This seems a little strange having session data on the `TemplateMgmtChoosePage` POM. 

Is it worth splitting out the session data into a session fixture? 

I'll add the suggestion to your ticket.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    baseURL,
+  }) => {
+    const chooseTemplatePage = new TemplateMgmtChoosePage(page);
+
+    await chooseTemplatePage.navigateToChooseTemplatePage(
+      TemplateMgmtChoosePage.emptySessionData.id
+    );
+
+    await expect(page).toHaveURL(
+      `${baseURL}/templates/choose-a-template-type/3d98b0c4-6666-0000-0000-95eb27590000`
```
Comment (bhansell1) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-choose-page.component.ts:16]: This is great!

I wonder if we could use the Id directly from the `sessionWithNoData` object so that it's a bit easier to tie this Id back to the scenario? 

Otherwise it just looks like a magic number (uuid) with no meaning.

File: package.json [hunk 14]
```diff
@@ -24,6 +24,7 @@
     "markdown-it": "^13.0.1",
     "mimetext": "^3.0.24",
     "next": "14.2.6",
+    "nhs-notify-web-template-management-ui-tests": "file:tests/test-team",
```
Comment (chris-elliott-nhsd) [inline package.json:27]: what's the benefit of having this here? it seems like it will install all the test-related stuff whenever we do an npm install, including for production builds, which i don't think is ideal

File: tests/test-team/README.md [hunk 15]
```diff
@@ -0,0 +1 @@
+# NHS Notify Template Management (WebUI) Automated Regression Tests
```
Comment (chris-elliott-nhsd) [inline tests/test-team/README.md:1]: this empty README doesn't seem very useful - either add to it or remove it

File: tests/test-team/config/playwright.config.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
+import { defineConfig } from '@playwright/test';
+
+export default defineConfig({
+  testDir: '../',
+  /* Run tests in files in parallel */
+  fullyParallel: true,
+  /* Fail the build on CI if you accidentally left test.only in the source code. */
+  forbidOnly: !!process.env.CI,
+  /* Retry on CI only */
+  retries: process.env.CI ? 2 : 0,
```
Comment (chris-elliott-nhsd) [inline tests/test-team/config/playwright.config.ts:10]: do we want to allow retries in CI? i know we do this in Notify core but would it be good to start off here committing to writing non-flaky tests?

File: tests/test-team/README.md (hunk 15 above)
Comment (bhansell1) [reply tests/test-team/README.md:1]: Might be a good place to talk about the difference between the e2e tests and component tests and where the boundary is between them?

File: tests/test-team/pages/template-mgmt-base-page.ts (hunk 6 above)
Comment (chris-elliott-nhsd) [reply tests/test-team/pages/template-mgmt-base-page.ts:19]: i agree with Sean here - i'm also not a fan of adding stuff to production code that's only for use by the tests

File: tests/test-team/config/playwright.config.ts (hunk 16 above)
Comment (chris-elliott-nhsd) [reply tests/test-team/config/playwright.config.ts]: updated

File: tests/test-team/README.md (hunk 15 above)
Comment (chris-elliott-nhsd) [reply tests/test-team/README.md:1]: added

File: tests/test-team/setup/session.setup.ts (hunk 12 above)
Comment (chris-elliott-nhsd) [reply tests/test-team/setup/session.setup.ts]: updated - test suites now seed their own data, none of this is global data so we don't really need a global setup at the moment

File: tests/test-team/package.json (hunk 11 above)
Comment (chris-elliott-nhsd) [reply tests/test-team/package.json]: updated

File: tests/test-team/template-mgmt-component-tests/template-mgmt-start-page.component.ts (hunk 10 above)
Comment (chris-elliott-nhsd) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-start-page.component.ts]: updated

**Review by bhansell1 — DISMISSED** [review]

File: package.json [hunk 17]
```diff
@@ -20,7 +20,12 @@
   "dependencies": {
     "@aws-amplify/adapter-nextjs": "^1.2.19",
     "@aws-amplify/backend": "^1.2.0",
+    "@aws-sdk/client-appsync": "^3.650.0",
+    "@aws-sdk/client-dynamodb": "^3.654.0",
```
Comment (mark-r-bjss) [inline package.json:24]: The tests should operate as their own app with a separate package.json so that we don't pollute the production code with unnecessary test code or in this case dependencies.

Comment (chris-elliott-nhsd) [reply package.json]: updated

File: .tool-versions [hunk 18]
```diff
@@ -4,6 +4,7 @@ pre-commit 3.6.0
 terraform 1.9.2
 vale 3.6.0
 tfsec 1.28.10
+nodejs 20.13.1
```
Comment (mark-r-bjss) [inline .tool-versions:7]: Do we also need to make sure that the backend lambda runtimes stays in sync with this?

File: package.json [hunk 19]
```diff
@@ -24,6 +24,7 @@
     "markdown-it": "^13.0.1",
     "mimetext": "^3.0.24",
     "next": "14.2.13",
+    "nhs-notify-web-template-management-ui-tests": "file:tests/test-team",
```
Comment (mark-r-bjss) [inline package.json:27]: Why do we need this?

File: .tool-versions (hunk 18 above)
Comment (chris-elliott-nhsd) [reply .tool-versions:7]: updated

File: package.json (hunk 19 above)
Comment (chris-elliott-nhsd) [reply package.json]: moved to dev dependency as discussed

**Review by mark-r-bjss — APPROVED** [review]

**Review by mark-r-bjss — APPROVED** [review]

**Review by jazrattanpal — APPROVED** [review]
Approved by Test

**Review by bhansell1 — DISMISSED** [review]

**Review by chris-elliott-nhsd — APPROVED** [review]

**Review by jazrattanpal — APPROVED** [review]

**Review by jazrattanpal — APPROVED** [review]

**Review by mark-r-bjss — APPROVED** [review]
